### ISSUE TITLE
CKEditor  - readOnly attribute is not accessible via customConfig

### ISSUE THREAD

**Issue by jxmai** [primefaces-extensions/primefaces-extensions.github.com#415]
The CKEditor shows a lot of grey buttons which are not usable in readonly mode:

![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19458486/ff4ab1d2-9491-11e6-8609-65c8777b859c.png)

**From user point of view, there is no point to show those buttons. They are simply waste of space.**

I suggest we can remove those buttons in read only mode, and we will end up like the following screenshot:

![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19458610/be7b1b96-9492-11e6-86db-d5d203124611.png)

---

Here is the important part:

Although I can configure the toolbar in customconfig, it will end up **impacting both readonly mode and not readonly mode**. It is better to make some changes in the primefaces core's ```widget.js```

In the widget.js, simply adding the following js will work:

```js

if(this.cfg.readOnly) {
            CKEDITOR.config.toolbar =  [
	     {name: 'tools', items: ['Source','Preview','Print','Find','Maximize']},
	     {name: 'about', items: ['About']}
	];
```

The exact location is:

![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19458722/57a12298-9493-11e6-9a2a-7c2d7d55e898.png)

**Comment by melloware** [comment]
@tandraschko thoughts on this?  I have no problem making the change but wanted your input first.

**Comment by tandraschko** [comment]
-1
we shoudln't change the default behavior of CKEditor

**Comment by jxmai** [comment]
@tandraschko @melloware I understand we shouldn't change the default bebavior of CKEditor, and I come up with another solution. There is another issue I found out.

In CKEditor documentation, there is a flag called ```config.readonly``` that can toggle the readonly mode (Here is the link: http://docs.ckeditor.com/#!/guide/dev_readonly)

I was expecting the following js in the ckeditor.js will work, but it did not

```js
CKEDITOR.editorConfig = function(config) {
	
	if(config.readOnly == true)
		{
		config.toolbar = [
		     {name: 'tools', items: ['Source','Preview','Print','Find','Maximize']},
		     {name: 'about', items: ['About']}
		];
	}
	
}
```

**The reason is config.readOnly is undefine ...**

Let me alert it:

```js
alert(config.readOnly);
```
![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19482348/6e6a0cdc-9516-11e6-9157-1ba1ea3fd8ba.png)


After further digging, I noticed that the readonly attribute in the primefaces component can actually overwrite the customconfig for some reason.

For example, if I define readonly = true in the component, and I also define readonly = false in the customconfig. The editor is in readonly mode. (The custom config should overwrite the previous setting but it did not)

In conclusion, Primefaces extension core use some other logic that is controlling the readOnly mode that cannot be read from the config file.

**Comment by tandraschko** [comment]
Nope, we pass it via options: https://github.com/primefaces-extensions/core/blob/master/src/main/resources/META-INF/resources/primefaces-extensions/ckeditor/widget.js#L97

Not sure why you can't access it via CKEDITOR.editorConfig = function(config)...

**Comment by melloware** [comment]
I can't see any reason why readOnly would be undefined at any point.

**Comment by jxmai** [comment]
The readOnly is not within config object:

Check this out: 

test.xhtml

```html
<ui:composition xmlns="http://www.w3.org/1999/xhtml"
	xmlns:h="http://java.sun.com/jsf/html"
	xmlns:p="http://primefaces.org/ui"
	xmlns:pe="http://primefaces.org/ui/extensions"
	xmlns:ui="http://java.sun.com/jsf/facelets"
	xmlns:f="http://java.sun.com/jsf/core"
	xmlns:fn="http://java.sun.com/jsp/jstl/functions"
	xmlns:i="http://java.sun.com/jsf/composite/components"
	xmlns:c="http://java.sun.com/jsp/jstl/core">

	<h:head>

	</h:head>

	<h:body>
		<h:form>
			<p:commandButton value="Load Data" 
						process="@this" 
						update="editor"></p:commandButton>
		
		
			<pe:ckEditor id="editor" 
						customConfig="#{request.contextPath}/resources/js/ckeditor.js"
						>
			</pe:ckEditor>
			
		</h:form>
	</h:body>

</ui:composition>
```

ckeditor.js

```js
CKEDITOR.editorConfig = function( config ) {
	alert(config.readOnly);
	console.log(config.readOnly);
};
```

The result is:

![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19494842/9095d5d6-9546-11e6-839c-ec949ab90d4f.png)


readOnly attribute is not in config

![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19494917/e1a652a2-9546-11e6-8a72-f32453f569d7.png)

**Comment by melloware** [comment]
Yep I confirmed it too.  Either their documentation is bad or we are not on the latest CKEditor?  Their sample code does this.  They set the editor instance to readonly.

`editor.setReadOnly( true);`

**Comment by jxmai** [comment]
This will work, but it is more like a patch...

```
CKEDITOR.config.readOnly = this.cfg.readOnly;
```

![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19495778/6e174e00-954a-11e6-8aa5-e32ccc5c66ea.png)

![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19495800/88512e3a-954a-11e6-9ee9-52c53150fd57.png)

![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19495863/b81f05ba-954a-11e6-884f-042a1760483d.png)


![snapshot2](https://cloud.githubusercontent.com/assets/10467831/19495892/da868b0a-954a-11e6-92d5-6ac2370cc8a4.png)

**Comment by melloware** [comment]
I don't think that is so bad. I will add it unless @tandraschko has any issues with this fix?

**Comment by tandraschko** [comment]
`CKEDITOR.config.readOnly = this.cfg.readOnly;`
will AFAICS break multiple editors on the page as `CKEDITOR.config` is static for all editors.

**Comment by melloware** [comment]
Good point.

**Comment by jxmai** [comment]
Good catch. That way won't work. Now I kind of understand the real problem:

Scene 1: When readOnly = true. **this.cfg = false**, the if block gets skipped, this.option.readOnly = undefined

![snapshot1](https://cloud.githubusercontent.com/assets/10467831/19545740/6b16fe6c-964e-11e6-9ca7-9454ebc97305.png)


Scene 2: When readOnly = false, this.cfg = false, the if block gets skipped, this.option.readOnly = undefined

---
Scene 2 is easy to solve, but the problem is the bold part. this.cfg should be true not false.

**Comment by jxmai** [comment]
Good catch. The previous method failed to detect the readonly state of multiple editors on the same page after my verification. 

Sample code is here (2 editors):

```xhtml
<ui:composition xmlns="http://www.w3.org/1999/xhtml"
     xmlns:h="http://xmlns.jcp.org/jsf/html"
    xmlns:p="http://primefaces.org/ui"
    xmlns:pe="http://primefaces.org/ui/extensions"
    xmlns:ui="http://xmlns.jcp.org/jsf/facelets"
    xmlns:f="http://xmlns.jcp.org/jsf/core"
    >

    <h:head>
    </h:head>

    <h:body>
        <h:form>
            <p:commandButton value="Load Data" 
                        process="@this" 
                        update="editor">
            </p:commandButton>


            <pe:ckEditor id="editor"  
                        customConfig="#{request.contextPath}/resources/js/ckeditor.js"
                        readonly="true"
            >
            </pe:ckEditor>


            <pe:ckEditor id="editor2"  
                        customConfig="#{request.contextPath}/resources/js/ckeditor.js"
                        readonly="false"
            >
            </pe:ckEditor>
        </h:form>
    </h:body>

 </ui:composition>
```

The way to fix this issue is to detect the current config instance and to make sure the instance does exist before assigning the readOnly attribute. 

 I raised a pull request to fix this issue: https://github.com/primefaces-extensions/core/pull/36

This way will work :)
